client/material/text-field.tsx

```diff
@@ -0,0 +1,373 @@
+    )
+
+    const onInputBlur = useCallback(
+      (event: React.FocusEvent<HTMLInputElement>) => {
+        setIsFocused(false)
+        onBlur?.(event)
+      },
+      [onBlur],
+    )
+
+    const onInputFocus = useCallback(
+      (event: React.FocusEvent<HTMLInputElement>) => {
+        setIsFocused(true)
+        onFocus?.(event)
+      },
+      [onFocus],
+    )
+
+    const onInputChange = useCallback(
+      (event: React.ChangeEvent<HTMLInputElement>) => {
+        if (multiline) {
+          autoSize(event.target)
+        }
+
+        onChange?.(event)
+      },
+      [autoSize, multiline, onChange],
+    )
+
+    const onInputKeyDown = useCallback(
+      (event: React.KeyboardEvent<HTMLInputElement>) => {
+        if (event.code === 'Enter' || event.code === 'EnterNumpad') {
+          onEnterKeyDown?.(event)
+        }
+
+        onKeyDown?.(event)
+      },
+      [onEnterKeyDown, onKeyDown],
+    )
+
+    const internalInputProps = {
+      id,
+      value,
+      type,
+      name,
+      disabled,
+      onBlur: onInputBlur,
+      onChange: onInputChange,
+      onFocus: onInputFocus,
+      onKeyDown: onInputKeyDown,
+    }
+
+    const leadingIconsElements = leadingIcons.map((leadingIcon, index) => {
+      const elem = React.cloneElement(leadingIcon, { tabIndex: -1 })
+      return (
+        <LeadingIcon key={index} $index={index} $dense={dense}>
+          {elem}
+        </LeadingIcon>
+      )
+    })
+    const trailingIconsElements = trailingIcons
+      .slice() // Don't mutate the original array
+      .reverse()
+      .map((trailingIcon, index) => {
+        const elem = React.cloneElement(trailingIcon, { tabIndex: -1 })
+        return (
+          <TrailingIcon key={index} $index={index} $dense={dense} $multiline={multiline}>
+            {elem}
+          </TrailingIcon>
+        )
+      })
+
+    const clearAndFocusInput = useStableCallback(() => {
+      if (!inputRef.current) {
+        return
+      }
+
+      // React overrides the input value setter so we have to do shenanigans to set a value
+      // manually. See this answer for more info: https://stackoverflow.com/a/46012210/398302
+      const nativeInputValue = Object.getOwnPropertyDescriptor(
+        window.HTMLInputElement.prototype,
+        'value',
+      )
+      nativeInputValue?.set?.call(inputRef.current, '')
+      const changeEvent = new Event('input', { bubbles: true })
+      inputRef.current.dispatchEvent(changeEvent)
+      inputRef.current?.focus()
+    })
+
+    if (hasClearButton && value) {
+      trailingIconsElements.push(
+        <TrailingIcon
+          key={trailingIconsElements.length}
+          $index={trailingIconsElements.length}
+          $dense={dense}
+          $multiline={multiline}>
+          <ClearTooltip text='Clear'>
+            <ClearButton icon={<ClearIcon />} onClick={clearAndFocusInput} />
+          </ClearTooltip>
+        </TrailingIcon>,
+      )
+    }
+
+    const renderLabel = () => {
```

Comment:
would prefer that this is just assigned to a variable rather than creating a function that we call and immediately throw away every time